File: servers/filesystem-sandbox/src/tools.ts
/**
 * Filesystem-sandbox MCP tool implementations.
 *
 * Each tool routes its `path` argument through `Sandbox.resolve(...)`
 * before any filesystem syscall. The sandbox layer is what makes the
 * tools safe; the tools themselves are thin glue.
 */

import { promises as fs } from "node:fs";
import path from "node:path";
import { Sandbox, SandboxEscape } from "./sandbox.js";

export interface ToolDeps {
  sandbox: Sandbox;
  readOnly: boolean;
  maxBytes: number;
}

export class WriteForbiddenError extends Error {
  constructor() {
    super("write_file is disabled (MCP_FS_SANDBOX_READ_ONLY=1)");
    this.name = "WriteForbiddenError";
  }
}

export class FileTooLargeError extends Error {
  constructor(public readonly size: number, public readonly limit: number) {
    super(`file size ${size} > limit ${limit} bytes`);
    this.name = "FileTooLargeError";
  }
}

export interface DirEntry {
  name: string;
  kind: "file" | "directory" | "symlink" | "other";
  size?: number;
}

export async function listDirectory(deps: ToolDeps, dir: string): Promise<DirEntry[]> {
  const sp = await deps.sandbox.resolveDir(dir);
  const entries = await fs.readdir(sp.resolved, { withFileTypes: true });
  const out: DirEntry[] = [];
  for (const e of entries) {
    let kind: DirEntry["kind"];
    if (e.isFile()) kind = "file";
    else if (e.isDirectory()) kind = "directory";
    else if (e.isSymbolicLink()) kind = "symlink";
    else kind = "other";
    const entry: DirEntry = { name: e.name, kind };
    if (kind === "file") {
      try {
        const stat = await fs.stat(path.join(sp.resolved, e.name));
        entry.size = stat.size;
      } catch {
        // ignore — stat can fail on broken symlinks; we already
        // know the entry type.
      }
    }
    out.push(entry);
  }
  out.sort((a, b) => a.name.localeCompare(b.name));
  return out;
}

export async function readFile(deps: ToolDeps, file: string): Promise<string> {
  const sp = await deps.sandbox.resolveFile(file);
  const stat = await fs.stat(sp.resolved);
  if (stat.size > deps.maxBytes) {
    throw new FileTooLargeError(stat.size, deps.maxBytes);
  }
  // Refuse anything that isn't UTF-8 text — binary files surface as
  // a clear error rather than as garbled bytes inside a JSON tool
  // result. Detection is "decode strict, fail on replacement char".
  const buf = await fs.readFile(sp.resolved);
  const decoder = new TextDecoder("utf-8", { fatal: true });
  try {
    return decoder.decode(buf);
  } catch {
    throw new Error(`file is not valid UTF-8 text: ${file}`);
  }
}

export async function writeFile(
  deps: ToolDeps,
  file: string,
  content: string,
): Promise<{ bytes_written: number }> {
  if (deps.readOnly) throw new WriteForbiddenError();
  const data = Buffer.from(content, "utf-8");
  if (data.byteLength > deps.maxBytes) {
    throw new FileTooLargeError(data.byteLength, deps.maxBytes);
  }
  // The file may not exist yet; resolve with `mustExist: false` so
  // the sandbox checks the parent's containment instead.
  const sp = await deps.sandbox.resolve(file, { mustExist: false });
  await fs.writeFile(sp.resolved, data);
  return { bytes_written: data.byteLength };
}

export function isSandboxEscape(err: unknown): err is SandboxEscape {
  return err instanceof SandboxEscape;
}

File: servers/filesystem-sandbox-py/filesystem_sandbox/tools.py
"""Filesystem-sandbox MCP tool implementations.

Each tool routes its ``path`` argument through ``Sandbox.resolve(...)``
before any filesystem syscall. The sandbox layer is what makes the
tools safe; the tools themselves are thin glue.

Mirrors ``../filesystem-sandbox/src/tools.ts`` shape-for-shape:
identical defensive posture, identical error classes, identical
output dicts.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Literal

from .sandbox import Sandbox

DirEntryKind = Literal["file", "directory", "symlink", "other"]


@dataclass
class ToolDeps:
    """Injected configuration the tool handlers need.

    Built once at server start; not mutated per call.
    """

    sandbox: Sandbox
    read_only: bool = False
    max_bytes: int = 1_048_576  # 1 MiB default; matches the TS server


class WriteForbiddenError(Exception):
    """Raised when ``write_file`` is called while ``read_only`` is True."""

    def __init__(self) -> None:
        super().__init__("write_file is disabled (MCP_FS_SANDBOX_READ_ONLY=1)")


class FileTooLargeError(Exception):
    """Raised when a file would exceed the configured byte cap."""

    def __init__(self, size: int, limit: int) -> None:
        super().__init__(f"file size {size} > limit {limit} bytes")
        self.size = size
        self.limit = limit


@dataclass(frozen=True)
class DirEntry:
    name: str
    kind: DirEntryKind
    size: int | None = None


def list_directory(deps: ToolDeps, dir_path: str) -> list[DirEntry]:
    """Return the directory contents as a sorted-by-name list of entries."""
    sp = deps.sandbox.resolve_dir(dir_path)
    out: list[DirEntry] = []
    for name in os.listdir(sp.resolved):
        full = os.path.join(sp.resolved, name)
        if os.path.islink(full):
            kind: DirEntryKind = "symlink"
            size: int | None = None
        elif os.path.isfile(full):
            kind = "file"
            try:
                size = os.path.getsize(full)
            except OSError:
                size = None
        elif os.path.isdir(full):
            kind = "directory"
            size = None
        else:
            kind = "other"
            size = None
        out.append(DirEntry(name=name, kind=kind, size=size))
    out.sort(key=lambda e: e.name)
    return out


def read_file(deps: ToolDeps, file_path: str) -> str:
    """Read the file as UTF-8 text. Caps at ``max_bytes``.

    Files that aren't valid UTF-8 raise a ``ValueError``; that's the
    same contract the TS server emits.
    """
    sp = deps.sandbox.resolve_file(file_path)
    size = os.path.getsize(sp.resolved)
    if size > deps.max_bytes:
        raise FileTooLargeError(size, deps.max_bytes)
    with open(sp.resolved, "rb") as fh:
        raw = fh.read()
    try:
        return raw.decode("utf-8", errors="strict")
    except UnicodeDecodeError as exc:
        raise ValueError(f"file is not valid UTF-8 text: {file_path}") from exc


def write_file(deps: ToolDeps, file_path: str, content: str) -> dict[str, int]:
    """Write ``content`` to ``file_path`` as UTF-8.

    Raises ``WriteForbiddenError`` when the server is read-only,
    ``FileTooLargeError`` when the content exceeds the cap, and any
    ``SandboxEscape`` from the sandbox layer.
    """
    if deps.read_only:
        raise WriteForbiddenError()
    data = content.encode("utf-8")
    if len(data) > deps.max_bytes:
        raise FileTooLargeError(len(data), deps.max_bytes)
    sp = deps.sandbox.resolve(file_path, must_exist=False)
    with open(sp.resolved, "wb") as fh:
        fh.write(data)
    return {"bytes_written": len(data)}
